### DISCUSSION TITLE
Using Minecraft types inside of an enumBuilder

### DISCUSSION THREAD

**Discussion by Desynq** [FabricMC/fabric-api#2351] [Mod Dev Support] [answered]
I'm currently trying to add an enum to `net.minecraft.class_189` (Advancement Frames) and currently am doing this:

```java
var remapper = FabricLoader.getInstance().getMappingResolver();
String advancementFrame = remapper.mapClassName("intermediary", "net.minecraft.class_189");

ClassTinkerers.enumBuilder(advancementFrame, String.class, Integer.class, Formatting.class).addEnum("OPTIONAL", "optional", 52, Formatting.BLUE).build();
```
Fabric-ASM, however, doesn't like that I'm referring to a Minecraft class in `enumBuilder()`, said class I'm referencing is `net.minecraft.class_124`, an enum that holds all the different formatting types.

I'm wondering if there's a way to reference the type in `net.minecraft.class_124` in `enumBuilder()` without it causing an error.

**Comment by warjort** [comment, marked as answer]
The docs have a whole section about what to avoid in the classloading: https://github.com/Chocohead/Fabric-ASM/blob/master/README.md#navigating-the-entrance
Basically, your shouldn't be loading any minecraft classes either directly, or via any of your own classes that reference/load minecraft classes. The processing occurs too early to start loading minecraft.

The example I linked on your previous question, shows one way to reference (but not load) a minecraft class for the constructor parameters. In that example it is the ItemStack class
https://github.com/Chocohead/Fabric-ASM/blob/master/example/src/com/chocohead/mm/testing/EarlyRiser.java#L22-L29

**Reply by Desynq** [reply]
Interesting, I'll try it out and see what happens, thanks.

**Reply by Desynq** [reply]
It seems to not be working properly when I try entering a world.

```java
[09:30:20] [Render thread/WARN] (Minecraft) Failed to load datapacks, can't proceed with server load
java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: java/lang/StringLjava/lang/IntegerLnet/minecraft/util/Formatting
```

My code currently is:
```java
public void run() {
  var remapper = FabricLoader.getInstance().getMappingResolver();
  String advancementFrame = remapper.mapClassName("intermediary", "net.minecraft.class_189");
  
  String formatting = 'L' + remapper.mapClassName("intermediary", "net.minecraft.class_124") + ';';
  ClassTinkerers.enumBuilder(advancementFrame, "Ljava/lang/String", "Ljava/lang/Integer", formatting).addEnum("STARTER", () -> new Object[] {"starter", 52, Formatting.BLUE}).build();
}
```

I'm wondering if I wrote the parameters wrong or something like that.

**Reply by warjort** [reply]
The jvm internal name of a class looks like this ````Lname;````
https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3.2-200
You are missing the semi-colon(s).

**Reply by Desynq** [reply]
Ok, got it, thanks.

**Reply by Desynq** [reply]
I changed it to this:
```java
ClassTinkerers.enumBuilder(advancementFrame, "Ljava/lang/String;", "Ljava/lang/Integer;", formatting).addEnum("STARTER", () -> new Object[] {"starter", 52, Formatting.BLUE}).build();
```

And now it's erroring with:
```java
java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: 'void net.minecraft.advancement.AdvancementFrame.<init>(java.lang.String, int, java.lang.String, java.lang.Integer, net.minecraft.util.Formatting)'
```

**Reply by warjort** [reply]
Please can you post all the relevant code and the exception's full stack trace.
I can only guess what the problem is from 1 line of code and an error with little context.

One thing I can say is the second parameter of the AchievementFrame constructor is an int, not a java.lang.Integer so the jvm internal class name will be "I".

Failing that, you will have to debug fabric-asm to see why it is not using the parameters you think it should be using.

**Reply by warjort** [reply]
I changed your code to this
````
public class EarlyRiser implements Runnable {
    @Override
    public void run() {
        MappingResolver remapper = FabricLoader.getInstance().getMappingResolver();
        String advancementFrame = remapper.mapClassName("intermediary", "net.minecraft.class_189");
        
        String formatting = 'L' + remapper.mapClassName("intermediary", "net.minecraft.class_124") + ';';
        ClassTinkerers.enumBuilder(advancementFrame, "Ljava/lang/String;", "I", formatting)
                      .addEnum("STARTER", () -> new Object[] {"starter", 52, Formatting.BLUE})
                      .build();
    }
}
````

Then added the following event handler to show the enum is there:
````
ServerLifecycleEvents.SERVER_STARTED.register(server -> LOGGER.info("AFs:" + Arrays.asList(AdvancementFrame.values())));
````
Which outputs
````
[Server thread/INFO] (modid) AFs:[TASK, CHALLENGE, GOAL, STARTER]
````

**Reply by Desynq** [reply]
Ok, thank you, it seems to be working now.
